File: 双指针/python/345.反转字符串中的元音字母.py
#
# @lc app=leetcode.cn id=345 lang=python3
#
# [345] 反转字符串中的元音字母
#

# @lc code=start
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowel = list("aeiouAEIOU")
        s_list = list(s)
        left, right = 0, len(s) - 1

        while left < right:
            while left < right and s_list[left] not in vowel:
               left = left + 1
               
            while left < right and s_list[right] not in vowel:
                right = right - 1

            s_list[left], s_list[right] = s_list[right], s_list[left]
            left = left + 1
            right = right - 1

        return ''.join(s_list)

# @lc code=end

File: 双指针/java/345.反转字符串中的元音字母.java
/*
 * @lc app=leetcode.cn id=345 lang=java
 *
 * [345] 反转字符串中的元音字母
 */

// @lc code=start
class Solution {
    public String reverseVowels(String s) {
        if (s == null || s.length() == 0) {
            return s;
        }

        int left = 0;
        int right = s.length() - 1;

        char[] chars = s.toCharArray(); 
        String vowel = "aeiouAEIOU";

        while (left < right) {
           while (left < right && !vowel.contains(chars[left] + "")) {
               left++;
           }
           while (left < right && !vowel.contains(chars[right] + "")) {
               right-- ;
           }
           swapChar(chars, left, right);
           left++;
           right--;
        }
        return new String(chars);
    }
    
    public void swapChar(char[] chars, int  left, int right) {
        char temp = chars[left];
        chars[left] = chars[right];
        chars[right] = temp;
    }
}
// @lc code=end

